src/replica/duplication/mutation_batch.cpp

```diff
@@ -20,13 +20,70 @@
 
 #include "replica_duplicator.h"
 #include "mutation_batch.h"
-#include "replica/prepare_list.h"
 
 namespace dsn {
 namespace replication {
 
 /*static*/ constexpr int64_t mutation_batch::PREPARE_LIST_NUM_ENTRIES;
 
+mutation_buffer::mutation_buffer(replica_base *r,
+                                 decree init_decree,
+                                 int max_count,
+                                 mutation_committer committer)
+    : prepare_list(r, init_decree, max_count, committer)
+{
+    auto counter_str = fmt::format("dup_recent_mutation_loss_count@{}", r->get_gpid());
+    _counter_dulication_mutation_loss_count.init_app_counter(
+        "eon.replica", counter_str.c_str(), COUNTER_TYPE_VOLATILE_NUMBER, counter_str.c_str());
+}
+
+void mutation_buffer::commit(decree d, commit_type ct)
+{
+    if (d <= last_committed_decree())
+        return;
+
+    if (ct != COMMIT_TO_DECREE_HARD) {
+        dassert_replica(false, "invalid commit type %d", (int)ct);
+    }
+
+    ballot last_bt = 0;
+    for (decree d0 = last_committed_decree() + 1; d0 <= d; d0++) {
+        mutation_ptr next_commit_mutation = get_mutation_by_decree(d0);
+        // The unexpected case as follow: next_commit_decree is out of [start~end]
+        //
+        // last_commit_decree - next_commit_decree
+        //                         |                                       |
+        //                        n                                    n+1
+        //
+        //  [min_decree------max_decree]
+        //                |                                |
+        //             n+m(m>1)            n+k(k>=m)
+        //
+        // just derror but not dassert if mutation loss or other problem, it's different from base
+        // class implement. And from the error and perf-counter, we can choose restart duplication
+        // or ignore the loss.
```

Comment:
In this comments, I suggest `last_commit_decree` -> `last_committed_decree`, `next_commit_decree` -> `next_committed_decree`, `next_commit_mutation` -> `next_committed_mutation`